doc/method/method.md

```diff
@@ -6,173 +6,348 @@ The assembly process works, in outline, as follows:
+improve matches between taxonomies.  Other patches such as extinct or
+extant annotations or topology changes can be applied at this point.
+Generally it's best to handle taxonomic problems by correcting source
+taxonomies at this point, but if it is difficult or inconvenient to
+localize the problem to one or more sources, repair can be left until
+the final patch phase after all source taxonomies have been loaded and
+merged.
+
+Editing the source taxonomy presents a provenance tracking problem
+that has yet to be addressed.
+
+## Align source to union
+
+It is important that source taxa be matched with union taxa when
+and only when this is appropriate.  A mistaken identity between a
+source taxon and a union taxon can be disastrous, leading not just to
+an incorrect tree but to downstream curation errors in OTU matching
+(e.g. putting a snail in flatworms).  A mistaken non-identity
+(separation) can also be a problem, since taxon duplication leads to
+incomplete annotation (information about one copy not propagating to
+the other) and to loss of unification opportunities in phylogeny
+synthesis.
+
+The process of matching source nodes with union nodes, or equivalently
+determining the identity of the corresponding taxa, is called
+"alignment" in the following.
+
+Ultimately there is no automatable test to determine whether alignment
+has been done correctly.  The process is necessarily heuristic.
+Difficult cases must be investigated manually and either repaired
+manually (patched) or repaired by improvements to the heuristics.
+
+Alignment proceeds one name at a time.  First, a list of candidate
+union matches is made for the source nodes having that name.  Then, a
+set of heuristics is applied to find a unique best union node match,
+if any, for each source node having the given name.
+
+[Going one name at a time is not really correct since it does not
+handle synonyms very well.  There is a kludge for dealing with
+synonyms, but it is not satisfactory.  An experimental version of
+the code deals with one source node at a time, with the potential for
+a different candidate set for each node having a given name, based on
+its synonyms.]
+
+Heuristics are of two kinds:
+
+ 1. Separation heuristics identify union nodes that mustn't be match
+    targets for the source node.
+ 2. Preference heuristics attempt to separate the set of candidate
+    targets into two groups, those that are more appropriate vs. less
+    appropriate as match targets for the given source node.
+
+A heuristic is essentially a two-place predicate, applied to a source
+node and a union node, that either succeeds (the two could match
+according to this heuristic) or fails (they do not match according to
+this heuristic).  The method for applying the heuristics is as
+follows:
+
+ 1. Start with a source node and a set C of union nodes (candidates).
+ 2. For each heuristic H:
+      1. Let C' = those members of C for which H succeeds.
```

Comment:
Yes, three loops, but you left out the heuristics loop. You iterate over
source nodes, and for each you iterate over heuristics, and for each
heuristic you iterate over the (remaining) candidates. I think that's what
the text says but I'll try to clarify.

* For every source node: compute the candidate union nodes for that node
    * For every heuristic: (attempt to reduce the size of the candidate
set:)
        * For every candidate: see whether the heuristic favors it or not